chain/chain/src/flat_storage_creator.rs

```diff
@@ -83,19 +171,170 @@ impl FlatStorageShardCreator {
+                            store_helper::remove_fetching_state_status(&mut store_update, shard_id);
+                            store_helper::start_catchup(&mut store_update, shard_id);
+                        }
+                        store_update.commit()?;
+
+                        Ok(())
+                    }
+                }
+            }
+            FlatStorageStateStatus::CatchingUp => {
+                let store = self.runtime_adapter.store();
+                let old_flat_head = store_helper::get_flat_head(store, shard_id).unwrap();
+                let mut flat_head = old_flat_head.clone();
+                let chain_final_head = chain_store.final_head()?;
+                let mut merged_delta = FlatStateDelta::default();
+
+                // Merge up to 50 deltas of the next blocks until we reach chain final head.
+                // TODO: consider merging 10 deltas at once to limit memory usage
+                for _ in 0..50 {
+                    let height = chain_store.get_block_height(&flat_head).unwrap();
+                    if height > chain_final_head.height {
+                        panic!("New flat head moved too far: new head = {flat_head}, height = {height}, final block height = {}", chain_final_head.height);
+                    }
+                    if height == chain_final_head.height {
+                        break;
+                    }
+                    flat_head = chain_store.get_next_block_hash(&flat_head).unwrap();
+                    let delta =
+                        store_helper::get_delta(store, shard_id, flat_head).unwrap().unwrap();
+                    merged_delta.merge(delta.as_ref());
+                }
+
+                if old_flat_head != flat_head {
+                    // If flat head changes, save all changes to store.
+                    let old_height = chain_store.get_block_height(&old_flat_head).unwrap();
+                    let height = chain_store.get_block_height(&flat_head).unwrap();
+                    debug!(target: "chain", %shard_id, %old_flat_head, %old_height, %flat_head, %height, "Catching up flat head");
+                    let mut store_update = self.runtime_adapter.store().store_update();
+                    store_helper::set_flat_head(&mut store_update, shard_id, &flat_head);
+                    merged_delta.apply_to_flat_state(&mut store_update);
+
+                    if height == chain_final_head.height {
+                        // If we reached chain final head, we can finish catchup and finally create flat storage.
+                        store_helper::finish_catchup(&mut store_update, shard_id);
+                        store_update.commit()?;
+                        debug!(target: "chain", %shard_id, %flat_head, %height, "Creating flat storage");
+                        assert_eq!(
+                            self.runtime_adapter.try_create_flat_storage_state_for_shard(
```

Comment:
nit: as a general practice, don't put something more than an evaluation statement inside assert. Let's move `self.runtime_adapter.try_create_flat_storage_state_for_shard` outside the assert and create a variable for the assert.